src/Kafka/Conduit/Combinator.hs

```diff
@@ -45,10 +45,9 @@ batchBy n = foldYield f g (0 :: Int, [])
     g (_, xs) = [reverse xs]
 
 batchByOrFlush :: Monad m => Int -> Conduit (Maybe a) m [a]
-batchByOrFlush n = do
-  ma <- join <$> await
-  case ma of
-    Just a -> do
-      yield [a]
-      batchByOrFlush n
-    Nothing -> return ()
+batchByOrFlush n = foldYield folder finish (0 :: Int, [])
+  where
+    folder Nothing  (_, xs)                 = ((0    ,   []), [reverse    xs ])
+    folder (Just a) (i, xs) | (i + 1) >= n  = ((0    ,   []), [reverse (a:xs)])
+    folder (Just a) (i, xs)                 = ((i + 1, a:xs),               [])
```

Comment:
Personal preference, but I'd prefer a second guard with `otherwise` over restating the pattern match without the guard, to make it obvious that that's the fall-through.